### ISSUE TITLE
SoS goes to single core at multiple dependent steps

### ISSUE THREAD

**Issue by gaow** [vatlab/SoS#972]
Sorry to keep posting here issues running large number of small jobs. I initially thought this is related to #971 but looks it is not. Basically I have a workflow like #971, but now I have 4 steps `A`, `B`, `C` and `D`. The complete sequences should be:

```
A -> B -> C
A -> B -> D
```

and I made it such that the default step depends on `C` and `D` which then depend on `A` and `B` respectively. In execution, I expect `A` and `B` be executed sequentially, then `C` and `D` in parallel. This should be a very intuitive outcome-oriented logic. 

But it was not the case. I do see `C` and `D` gets started the same time, but only `C` gets executed in parallel. `D` drags along with only one thread, until it finishes all the computation. This is weird.  I've got a MWE here:

[16-May-2018-drag.zip](https://github.com/vatlab/SoS/files/2009342/16-May-2018-drag.zip)

This example has a slightly none-trivial step `A` but the rest steps are trivial. 

To reproduce the drag, first open `top` in another terminal, then

```
time sos run step_drag.sos -j 20
```

you'll see all seems fine (nicely parallelled) until steps `score_pi0_0` and `score_beta_0` (my `C` and `D`) above. One of them finished quickly but the other was dragging it running single process computation, although I would expect by logic that `C` and `D` should complete at the same time! The time it took was:

```
real	1m22.695s
```

Now, run the other example that removes step `D` only keep step `C`, 

```
time sos run step_no_drag.sos -s force -j 20
```

everything is paralleled and time took is only:

```
real	0m21.712s
```

So something strange is happening with this outcome-oriented style execution. The DAG file [drag_test.pdf](https://github.com/vatlab/SoS/files/2009322/drag_test.pdf)
 looks good.  File lock is my only guess now. In my MWE zip file I also put in a version called `step_drag_task_version.sos` that removes all `concurrent = True` and replace with `task: ` for actions. Running that version, it seems `C` and `D` completed at the same time!

**Comment by BoPeng** [comment]
Without running your example, I suspect that this is what happens:

1. C and D starts
2. C got all the remaining "slots", D got the one for the worker
3. C completes, D continues to use the one slot that was assigned

**Comment by gaow** [comment]
I agree with your suspicion, because that is the simplest theory to explain what is going on!

**Comment by BoPeng** [comment]
Note that sharing pools are not possible because the pools are created with processes that are forked from the worker process for each step. The only way to make it more clever is to checking for more resources and expand dynamically during the lifetime of a pool. I do not know how difficult it is to do this.

**Comment by BoPeng** [comment]
I do not see a need for the `shrink` scenario because there is no need to shrink as long as all computing resources (processes) are busy. I am checking if it is possible for a pool to expand if there are remaining resources.

**Comment by gaow** [comment]
>  I am checking if it is possible for a pool to expand if there are remaining resources.

Thanks a lot, that would be of tremendously useful!

**Comment by BoPeng** [comment]
The celery fork of multiprocessing, `billiard`, provides an enhanced pool with a `grow` function. I have not tested it heavily and have not made `billiard` the default, but if you install `billiard` locally, sos should now be able to grow the pool when there are slots available.

**Comment by gaow** [comment]
Great! I installed `billiard` locally, and I think it worked -- because it reduced my toy benchmark from 180 sec to 72 sec, and I can tell it is paralleled almost all the time.

So what's the deal here? Should we make it default, or maybe we need more tests? Or do we expect Python's multiprocessing will later catch it up?

**Comment by BoPeng** [comment]
As far as I remember, `billiard` was used a while ago but it caused some compatibility problem. Although here the usage is limited within a step and should be ok. I will let you use it for a while and add it to setup.py if no adverse effect is observed.

**Comment by gaow** [comment]
I recall having used `celery` some time ago that completely replaced Python's default `multiprocessing`, but indeed we switched it back ... here we are only using the `Pool` feature so hopefully it works! I'll close this issue for now until we have a problem.

**Comment by gaow** [comment]
So I searched for `billiard` in git history just out of curiosity and found [this patch](https://github.com/vatlab/SoS/commit/e46f5c728d007ccc45e7a9c10f7d8efeb34600ff) and [this issue](https://github.com/vatlab/SoS/issues/71) which I believe was the reason we got rid of it. So I updated to the latest `sklearn` (`pip install -U scikit-learn`) and ran this:

```
from sklearn import manifold
[1]
test = [1,2,3]
input: for_each = 'test', concurrent = True
from sklearn import manifold
print(9)
```

I had no problem. So we should be good, even to replace it globally?

But back to the patch that [introduced billiard](https://github.com/vatlab/SoS/commit/c30190a2a3c09dd3e39b02c1f0125e1fb4b2fd7b) it seems there was some limitations of the stock multiprocessing you want to overcome. I take that later billiard was not needed because you've got a better solution.
